posteriors/utils.py

```diff
@@ -228,6 +230,137 @@ def fisher(*args, **kwargs):
     return fisher
 
 
+def _vdot_real_part(x: Tensor, y: Tensor) -> float:
+    """Vector dot-product guaranteed to have a real valued result despite
+    possibly complex input. Thus neglects the real-imaginary cross-terms.
+
+    Args:
+        x: First tensor in the dot product.
+        y: Second tensor in the dot product.
+
+    Returns:
+        The result vector dot-product, a real float
+    """
+    # all our uses of vdot() in CG are for computing an operator of the form
+    #  z^H M z
+    #  where M is positive definite and Hermitian, so the result is
+    # real valued:
+    # https://en.wikipedia.org/wiki/Definiteness_of_a_matrix#Definitions_for_complex_matrices
+    real_part = torch.einsum("...i,...i->...", x.real, y.real)
```

Comment:
why `einsum` and not `torch.vdot`?

Reply:
torch vdot doesn't work with arrays that are more than 1D (see note here: https://pytorch.org/docs/stable/generated/torch.vdot.html)